### PR TITLE
"not for commit", handwritten word recognition for iam dataset

### PR DESCRIPTION
**Pull request by aarora8**
This branch is for handwritten word recognition on text line images. It uses word language model.

- add scripts for data preparation (text, wav.scp and utt2spk file) (local/prepare_data.sh, local/process_data.py)
- add scripts for feature extraction (local/make_feature_vect.py)
- add scripts for lexicon, language modeling, grammar (egs/iam/s5/local/prepare_lmWoOOV.py, egs/iam/s5/local/prepare_lm.sh, egs/iam/s5/local/prepare_lexicon.py, egs/iam/s5/local/prepare_dict.sh, egs/iam/s5/local/make_isolated_word_grammar.py, egs/iam/s5/local/add_test_data_to_train.py)
- add script for GMM-HMM training and using chain model (egs/iam/s5/local/chain/.run_cnn_1b.sh,  egs/iam/s5/local/chain/align_nnet3_lats.sh,  egs/iam/s5/run.sh)
- modification in nnet3/decode.sh to make it compatible with  egs/iam/s5/local/chain/.run_cnn_1b.sh
- add changes for variance floor + src changes (src/gmmbin/gmm-est.cc, src/gmm/mle-diag-gmm.h, egs/wsj/s5/steps/train_mono.sh, egs/iam/s5/run.sh)
- add modification for using 2 state HMM model (egs/wsj/s5/utils/gen_topo.pl)
- updated number of realign iterations (egs/wsj/s5/steps/train_sat.sh, egs/wsj/s5/steps/train_mono.sh, egs/wsj/s5/steps/train_lda_mllt.sh, egs/wsj/s5/steps/train_deltas.sh)

@hhadian @ChunChiehChang @YiwenShaoStephen

### CODE REVIEW

File: egs/wsj/s5/steps/train_mono.sh [hunk 1]
```diff
@@ -22,6 +22,7 @@ stage=-4
 power=0.25 # exponent to determine number of gaussians from occurrence counts
 norm_vars=false # deprecated, prefer --cmvn-opts "--norm-vars=false"
 cmvn_opts=  # can be used to add extra options to cmvn.
+variance_floor_val=0
```
Comment (danpovey) [inline egs/wsj/s5/steps/train_mono.sh]: should be variance_floor, I think, not variance_floor_val.

File: egs/wsj/s5/steps/train_lda_mllt.sh [hunk 2]
```diff
@@ -15,8 +15,8 @@ cmd=run.pl
 config=
 stage=-5
 scale_opts="--transition-scale=1.0 --acoustic-scale=0.1 --self-loop-scale=0.1"
-realign_iters="10 20 30";
-mllt_iters="2 4 6 12";
+realign_iters="2 5 10 15 20 25 30";
```
Comment (danpovey) [inline egs/wsj/s5/steps/train_lda_mllt.sh]: you shouldn't be changing defaults in this shared script.

File: src/gmm/mle-diag-gmm.h [hunk 3]
```diff
@@ -66,6 +70,14 @@ struct MleDiagGmmOptions {
                  module+"Variance floor (absolute variance).");
     opts->Register("remove-low-count-gaussians", &remove_low_count_gaussians,
                  module+"If true, remove Gaussians that fall below the floors.");
+    opts->Register("variance-floor-value", &variance_floor_value,
+                 module+"Value assigned to all elements of variance_floor_vector.");
+    opts->Register("variance-floor-vector-dim", &VFV_dim,
+                 module+"dimension of variance_floor_vector.");
+  }
+  void create_VFV() {
```
Comment (danpovey) [inline src/gmm/mle-diag-gmm.h]: this code doesn't have the right style (study Google style).  I suspect it's not needed anyway for the overal commit.

**Comment by danpovey** [comment]
@hhadian, please review this if it's to be committed.
I notice there are capitals in a python-script name; please remove them if
this is for commit.

**Comment by aarora8** [comment]
Thanks, updated title in pull request.

**Comment by aarora8** [comment]
removed unnecessary files, updated default value in shared script, removed changes for variance floor.

File: egs/iam/s5/run.sh [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+boost_sil=1
+beam=20
+
+lang_other=${num_states}
+mono_other=${lang_other}_${num_gauss}_var_beam${beam}_boost${boost_sil}
+tri_other=${mono_other}_${numLeavesTri}_${numGaussTri}
+tri2_other=${tri_other}_${numLeavesMLLT}_${numGaussMLLT}
+tri3_other=${tri2_other}_${numLeavesSAT}_${numGaussSAT}
+
```
Comment (hhadian) [inline egs/iam/s5/run.sh]: I guess it's better to change `other` to `affix` in these variables.

**Pull request closed without merging**